### PR TITLE
bugfix: SharePoint permissions fetching should be opt-in

### PR DESCRIPTION
**Pull request by ahmetmeleq**
Closes: #1891 (check the issue for more info)

### CODE REVIEW

**Comment by ahmetmeleq** [comment]
Came up to this in #1774 and had a quick fix already

Now adding opt-in functionality 👍

**Comment by ahmetmeleq** [comment]
Added a second test without permissions to test the problem in the future. 

However for the permission-test to work (and produce a different result than the test w/o permissions), docs need to be set a few permissions in Sharepoint platform itself

**Review by ryannikolaidis — CHANGES_REQUESTED** [review]
tested with:

```
PYTHONPATH=. ./unstructured/ingest/main.py \
  sharepoint \
  --download-dir /tmp/sharepoint-demo/download \
  --num-processes 6 \
  --strategy fast \
  --preserve-downloads \
  --reprocess \
  --output-dir /tmp/sharepoint-demo/result \
  --verbose \
  --client-cred "$SHAREPOINT_CRED" \
  --client-id "$SHAREPOINT_CLIENT_ID" \
  --site "$SHAREPOINT_SITE" \
  --path "Shared Documents" \
  --recursive
  ```
  
  hit error:
  
  ```
  
2023-10-26 14:04:40,230 MainProcess DEBUG    options: {'download_dir': '/tmp/sharepoint-demo/download', 'num_processes': 6, 'strategy': 'fast', 'preserve_downloads': True, 'reprocess': True, 'output_dir': '/tmp/sharepoint-demo/result', 'verbose': True, 'client_cred': '*******', 'client_id': '*******', 'site': '', 'path': 'Shared Documents', 'recursive': True, 'skip_infer_table_types': None, 'pdf_infer_table_structure': False, 'ocr_languages': None, 'encoding': None, 'fields_include': ['element_id', 'text', 'type', 'metadata', 'embeddings'], 'flatten_metadata': False, 'metadata_include': [], 'metadata_exclude': [], 'partition_by_api': False, 'partition_endpoint': 'https://api.unstructured.io/general/v0/general', 'api_key': None, 're_download': False, 'download_only': False, 'max_docs': None, 'embedding_api_key': None, 'embedding_model_name': None, 'chunk_elements': False, 'chunk_multipage_sections': False, 'chunk_combine_under_n_chars': 500, 'chunk_new_after_n_chars': 1500, 'work_dir': '/Users/ryannikolaidis/.cache/unstructured/ingest/pipeline', 'raise_on_error': False, 'permissions_application_id': None, 'permissions_client_cred': None, 'permissions_tenant': None, 'max_retries': None, 'max_retry_time': None, 'files_only': False}
2023-10-26 14:04:40,231 MainProcess ERROR    Please provide one of the following mandatory values:
--client-id
--client-cred
--site
Traceback (most recent call last):
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/cli/base/src.py", line 41, in src
    runner.run(**options)
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/runner/sharepoint.py", line 51, in run
    connector_config=SimpleSharepointConfig(
  File "<string>", line 10, in __init__
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/connector/sharepoint.py", line 45, in __post_init__
    raise ValueError(
ValueError: Please provide one of the following mandatory values:
--client-id
--client-cred
--site
Error: Please provide one of the following mandatory values:
--client-id
--client-cred
--site
```

**Comment by ahmetmeleq** [comment]
@ryannikolaidis , I cannot reproduce, I suspect if env variables were defined on your run

Here're my outputs with a copy-paste from your command:

```
% PYTHONPATH=. ./unstructured/ingest/main.py \
  sharepoint \
  --download-dir /tmp/sharepoint-demo/download \
  --num-processes 6 \
  --strategy fast \
  --preserve-downloads \
  --reprocess \
  --output-dir /tmp/sharepoint-demo/result \
  --verbose \
  --client-cred "$SHAREPOINT_CRED" \
  --client-id "$SHAREPOINT_CLIENT_ID" \
  --site "$SHAREPOINT_SITE" \
  --path "Shared Documents" \
  --recursive

2023-10-27 11:10:41,625 MainProcess DEBUG    .........
2023-10-27 11:10:41,626 MainProcess INFO     running pipeline: DocFactory -> Reader -> Partitioner -> Copier with config: .........
2023-10-27 11:10:42,857 MainProcess INFO     Running doc factory to generate ingest docs. Source connector: {......... "connector_config": {... "permissions_config": {"application_id": null, "client_cred": null, "tenant": null}}}
2023-10-27 11:10:42,973 MainProcess INFO     Permissions config is not fully fed with 'application_id', 'client_cred', and 'tenant'.                        Skipping permissions ingestion.
2023-10-27 11:10:44,320 MainProcess INFO     processing 5 docs via 6 processes
.........
2023-10-27 11:10:50,513 SpawnPoolWorker-12 INFO     writing partitioned content to /Users/ahmet/.cache/unstructured/ingest/pipeline/partitioned/d76fffa6a871331e83217aa3deb5525a.json
2023-10-27 11:10:50,530 MainProcess INFO     Calling Copier with 5 docs
.........
```

**Comment by ahmetmeleq** [comment]
Updated Sharepoint ingest test org for new permissions

**Comment by ahmetmeleq** [comment]
Running an ingest test fixtures update PR

**Review by ryannikolaidis — APPROVED** [review]
hmm, maybe I wasn't merged up? I just did and ran again (no changes otherwise) and it worked! lgtm!

File: unstructured/ingest/connector/sharepoint.py [hunk 1]
```diff
@@ -373,7 +373,18 @@ def initialize(self):
     def get_ingest_docs(self):
         base_site_client = self.connector_config.get_site_client()
 
-        if self.connector_config.permissions_config:
+        if not all(
+            [  # noqa: C419
+                value
+                for key, value in self.connector_config.permissions_config.to_dict().items()
+                if key in ["application_id", "client_cred", "tenant"]
```
Comment (badGarnet) [inline unstructured/ingest/connector/sharepoint.py:380]: can't we just do `getattr(self.connector_config.permissions_config, attr) for attr in ["application_id", "client_cred", "tenant"]`

Comment (badGarnet) [reply unstructured/ingest/connector/sharepoint.py:380]: that way we avoid running extra dict conversion

Comment (ahmetmeleq) [reply unstructured/ingest/connector/sharepoint.py]: Agreed, added with https://github.com/Unstructured-IO/unstructured/pull/1894/commits/0dc1a981854be4e69ee76e0e850c090a37ab49ed

**Pull request merged**